File: app/src/main/java/uk/co/sundroid/activity/data/fragments/YearEventsFragment.java
package uk.co.sundroid.activity.data.fragments;

import android.content.Intent;
import android.net.Uri;
import android.os.Handler;
import android.text.Html;
import android.view.View;
import android.view.ViewGroup;
import uk.co.sundroid.R.id;
import uk.co.sundroid.R.layout;
import uk.co.sundroid.activity.data.fragments.dialogs.settings.YearEventsPickerFragment;
import uk.co.sundroid.util.StringUtils;
import uk.co.sundroid.util.astro.MoonPhaseEvent;
import uk.co.sundroid.util.astro.SunDay;
import uk.co.sundroid.util.astro.math.MoonPhaseCalculator;
import uk.co.sundroid.util.astro.math.SunCalculator;
import uk.co.sundroid.domain.LocationDetails;
import uk.co.sundroid.util.prefs.SharedPrefsHelper;
import uk.co.sundroid.util.theme.ThemePalette;
import uk.co.sundroid.util.time.TimeUtils;
import uk.co.sundroid.util.time.Time;
import uk.co.sundroid.util.astro.YearData;
import uk.co.sundroid.util.astro.YearData.Event;
import uk.co.sundroid.util.astro.YearData.EventType;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.Set;

public class YearEventsFragment extends AbstractYearFragment implements ConfigurableFragment {

    private final Handler handler = new Handler();

	@Override
	protected int getLayout() {
		return layout.frag_data_yearevents;
	}

    @Override
    public void openSettingsDialog() {
//        boolean[] currentEvents = new boolean[8];
//        currentEvents[0] = SharedPrefsHelper.INSTANCE.getShowElement(getApplicationContext(), "yearNewMoon", true);
//        currentEvents[1] = SharedPrefsHelper.INSTANCE.getShowElement(getApplicationContext(), "yearFullMoon", true);
//        currentEvents[2] = SharedPrefsHelper.INSTANCE.getShowElement(getApplicationContext(), "yearQuarterMoon", true);
//        currentEvents[3] = SharedPrefsHelper.INSTANCE.getShowElement(getApplicationContext(), "yearSolstice", true);
//        currentEvents[4] = SharedPrefsHelper.INSTANCE.getShowElement(getApplicationContext(), "yearEquinox", true);
//        currentEvents[5] = SharedPrefsHelper.INSTANCE.getShowElement(getApplicationContext(), "yearLunarEclipse", true);
//        currentEvents[6] = SharedPrefsHelper.INSTANCE.getShowElement(getApplicationContext(), "yearSolarEclipse", true);
//        currentEvents[7] = SharedPrefsHelper.INSTANCE.getShowElement(getApplicationContext(), "yearEarthApsis", true);

        YearEventsPickerFragment settingsDialog = YearEventsPickerFragment.newInstance();
        settingsDialog.setTargetFragment(this, 0);
        settingsDialog.show(getFragmentManager(), "yearEventsSettings");
    }

    @Override
	protected void update(final LocationDetails location, final Calendar calendar, final View view) throws Exception {

        Thread thread = new Thread() {
            @Override
            public void run() {
                if (!isSafe()) {
                    return;
                }

                final Calendar todayCalendar = Calendar.getInstance(calendar.getTimeZone());
                final Set<Event> eventsSet = YearData.INSTANCE.getYearEvents(calendar.get(Calendar.YEAR), location.getTimeZone().getZone());
                List<MoonPhaseEvent> moonPhases = MoonPhaseCalculator.INSTANCE.getYearEvents(calendar.get(Calendar.YEAR), location.getTimeZone().getZone());
                for (MoonPhaseEvent moonPhase : moonPhases) {
                    eventsSet.add(new Event(EventType.PHASE, moonPhase, moonPhase.getTime(), null));
                }

                handler.post(() -> {
                    if (!isSafe()) {
                        return;
                    }

                    List<Event> eventsList = new ArrayList<>(eventsSet);
                    ViewGroup eventsBox = view.findViewById(id.yearEventsBox);
                    eventsBox.removeAllViews();

                    boolean first = true;
                    for (Event event : eventsList) {
                        Time eventTime = TimeUtils.formatTime(getApplicationContext(), event.getTime(), false);
                        String title = "";
                        String time = eventTime.getTime() + eventTime.getMarker().toLowerCase();
                        String subtitle = "";
                        String link = null;
                        int image = 0;
                        switch (event.getType()) {
                            case EARTH_APHELION:
                            case EARTH_PERIHELION:
                                if (!SharedPrefsHelper.INSTANCE.getShowElement(getApplicationContext(), "yearEarthApsis", true)) { continue; }
                                title = event.getType().getDisplayName();
                                link = event.getLink();
                                break;
                            case PARTIAL_LUNAR:
                            case TOTAL_LUNAR:
                            case PENUMBRAL_LUNAR:
                                if (!SharedPrefsHelper.INSTANCE.getShowElement(getApplicationContext(), "yearLunarEclipse", true)) { continue; }
                                title = event.getType().getDisplayName();
                                time = "Greatest eclipse: " + time;
                                link = event.getLink();
                                break;
                            case PARTIAL_SOLAR:
                            case TOTAL_SOLAR:
                            case ANNULAR_SOLAR:
                            case HYBRID_SOLAR:
                                if (!SharedPrefsHelper.INSTANCE.getShowElement(getApplicationContext(), "yearSolarEclipse", true)) { continue; }
                                title = event.getType().getDisplayName();
                                time = "Greatest eclipse: " + time;
                                subtitle = (String) event.getExtra();
                                link = event.getLink();
                                break;
                            case MARCH_EQUINOX:
                            case SEPTEMBER_EQUINOX:
                                if (!SharedPrefsHelper.INSTANCE.getShowElement(getApplicationContext(), "yearEquinox", true)) { continue; }
                                title = event.getType().getDisplayName();
                                break;
                            case NORTHERN_SOLSTICE:
                            {
                                if (!SharedPrefsHelper.INSTANCE.getShowElement(getApplicationContext(), "yearSolstice", true)) { continue; }
                                title = event.getType().getDisplayName();
                                if (Math.abs(location.getLocation().getLatitude().getDoubleValue()) > 23.44) {
                                    SunDay sunDay = SunCalculator.INSTANCE.calcDay(location.getLocation(), event.getTime(), SunCalculator.Event.RISESET);
                                    String localExtreme = location.getLocation().getLatitude().getDoubleValue() >= 0 ? "Longest" : "Shortest";
                                    subtitle = localExtreme + " day: " + TimeUtils.formatDurationHMS(getApplicationContext(), sunDay.getUptimeHours(), true);
                                }
                                break;
                            }
                            case SOUTHERN_SOLSTICE:
                            {
                                if (!SharedPrefsHelper.INSTANCE.getShowElement(getApplicationContext(), "yearSolstice", true)) { continue; }
                                title = event.getType().getDisplayName();
                                if (Math.abs(location.getLocation().getLatitude().getDoubleValue()) > 23.44) {
                                    SunDay sunDay = SunCalculator.INSTANCE.calcDay(location.getLocation(), event.getTime(), SunCalculator.Event.RISESET);
                                    String localExtreme = location.getLocation().getLatitude().getDoubleValue() >= 0 ? "Shortest" : "Longest";
                                    subtitle = localExtreme + " day: " + TimeUtils.formatDurationHMS(getApplicationContext(), sunDay.getUptimeHours(), true);
                                }
                                break;
                            }
                            case PHASE:
                                MoonPhaseEvent moonPhase = (MoonPhaseEvent) event.getExtra();
                                switch (moonPhase.getPhase()) {
                                    case FULL:
                                        if (!SharedPrefsHelper.INSTANCE.getShowElement(getApplicationContext(), "yearFullMoon", true)) { continue; }
                                        title = "Full Moon";
                                        image = ThemePalette.getPhaseFull();
                                        break;
                                    case NEW:
                                        if (!SharedPrefsHelper.INSTANCE.getShowElement(getApplicationContext(), "yearNewMoon", true)) { continue; }
                                        title = "New Moon";
                                        image = ThemePalette.getPhaseNew();
                                        break;
                                    case FIRST_QUARTER:
                                        if (!SharedPrefsHelper.INSTANCE.getShowElement(getApplicationContext(), "yearQuarterMoon", true)) { continue; }
                                        title = "First Quarter";
                                        image = location.getLocation().getLatitude().getDoubleValue() >= 0 ? ThemePalette.getPhaseRight() : ThemePalette.getPhaseLeft();
                                        break;
                                    case LAST_QUARTER:
                                        if (!SharedPrefsHelper.INSTANCE.getShowElement(getApplicationContext(), "yearQuarterMoon", true)) { continue; }
                                        title = "Last Quarter";
                                        image = location.getLocation().getLatitude().getDoubleValue() >= 0 ? ThemePalette.getPhaseLeft() : ThemePalette.getPhaseRight();
                                        break;
                                }
                                break;


                        }

                        if (!first) {
                            getActivity().getLayoutInflater().inflate(layout.divider, eventsBox);
                        }
                        View eventRow = View.inflate(getActivity(), layout.frag_data_yearevents_event, null);

                        boolean today = todayCalendar.get(Calendar.YEAR) == event.getTime().get(Calendar.YEAR) &&
                                todayCalendar.get(Calendar.MONTH) == event.getTime().get(Calendar.MONTH) &&
                                todayCalendar.get(Calendar.DAY_OF_MONTH) == event.getTime().get(Calendar.DAY_OF_MONTH);
                        if (today) {
                            eventRow.setBackgroundColor(ThemePalette.getCalendarHighlightColor());
                        } else {
                            eventRow.setBackgroundColor(ThemePalette.getCalendarDefaultColor());
                        }

                        if (image > 0) {
                            imageInView(eventRow, id.yearEventImg, image);
                            showInView(eventRow, id.yearEventImg);
                        }
                        textInView(eventRow, id.yearEventDate, Integer.toString(event.getTime().get(Calendar.DAY_OF_MONTH)));
                        textInView(eventRow, id.yearEventMonth, getShortMonth(event.getTime()));
                        textInView(eventRow, id.yearEventTitle, Html.fromHtml(title));
                        textInView(eventRow, id.yearEventTime, Html.fromHtml(time));
                        if (StringUtils.isNotEmpty(subtitle)) {
                            textInView(eventRow, id.yearEventSubtitle, Html.fromHtml(subtitle));
                            showInView(eventRow, id.yearEventSubtitle);
                        }
                        if (StringUtils.isNotEmpty(link)) {
                            final String finalLink = link;
                            showInView(eventRow, id.yearEventLink);
                            eventRow.setOnClickListener(view1 -> {
                                Intent intent = new Intent(Intent.ACTION_VIEW);
                                intent.setData(Uri.parse(finalLink));
                                startActivity(intent);
                            });
                        } else {
                            eventRow.setClickable(false);
                            eventRow.setFocusable(false);
                        }

                        eventsBox.addView(eventRow);
                        first = false;
                    }
                    if (first) {
                        removeInView(view, id.yearEventsBox);
                        showInView(view, id.yearEventsNone);
                    } else {
                        showInView(view, id.yearEventsBox);
                        removeInView(view, id.yearEventsNone);
                    }
                });
            }
        };
        thread.start();
    }

    private String getShortMonth(Calendar calendar) {
        switch (calendar.get(Calendar.MONTH)) {
            case 0: return "JAN";
            case 1: return "FEB";
            case 2: return "MAR";
            case 3: return "APR";
            case 4: return "MAY";
            case 5: return "JUN";
            case 6: return "JUL";
            case 7: return "AUG";
            case 8: return "SEP";
            case 9: return "OCT";
            case 10: return "NOV";
            case 11: return "DEC";
        }
        return "";

    }

}

File: app/src/main/kotlin/uk/co/sundroid/activity/data/fragments/YearEventsFragment.kt
package uk.co.sundroid.activity.data.fragments

import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.ViewGroup
import uk.co.sundroid.R
import uk.co.sundroid.R.id.*
import uk.co.sundroid.activity.MainActivity
import uk.co.sundroid.activity.data.fragments.dialogs.settings.YearEventsPickerFragment
import uk.co.sundroid.util.*
import uk.co.sundroid.util.astro.MoonPhaseEvent
import uk.co.sundroid.util.astro.math.MoonPhaseCalculator
import uk.co.sundroid.util.astro.math.SunCalculator
import uk.co.sundroid.domain.LocationDetails
import uk.co.sundroid.util.astro.MoonPhase
import uk.co.sundroid.util.prefs.Prefs
import uk.co.sundroid.util.theme.*
import uk.co.sundroid.util.time.*
import uk.co.sundroid.util.astro.YearData
import uk.co.sundroid.util.astro.YearData.Event
import uk.co.sundroid.util.astro.YearData.EventType
import uk.co.sundroid.util.async.async
import java.util.*
import java.util.Calendar.*
import kotlin.math.abs

class YearEventsFragment : AbstractYearFragment() {

    override val layout: Int
        get() = R.layout.frag_data_yearevents

    fun openSettingsDialog() {
        val settingsDialog = YearEventsPickerFragment.newInstance(requireContext())
        settingsDialog.setTargetFragment(this, 0)
        settingsDialog.show(requireFragmentManager(), "yearEventsSettings")
    }

    override fun onResume() {
        super.onResume()
        (activity as MainActivity).apply {
            setToolbarSubtitle(R.string.data_year_title)
            setViewConfigurationCallback({ openSettingsDialog() })
        }
    }

    @Throws(Exception::class)
    override fun update(location: LocationDetails, calendar: Calendar, view: View) {

        async(
                inBackground = {
                    val eventsSet = YearData.getYearEvents(calendar.get(YEAR), location.timeZone!!.zone)
                    val moonPhases = MoonPhaseCalculator.getYearEvents(calendar.get(YEAR), location.timeZone!!.zone)
                    moonPhases.mapTo(eventsSet) { Event(EventType.PHASE, it, it.time, null) }
                    eventsSet
                },
                onDone = { eventsSet: Set<Event> ->
                    if (isSafe) {
                        val todayCalendar = getInstance(calendar.timeZone)
                        val eventsList = ArrayList(eventsSet)
                        val eventsBox = view.findViewById<ViewGroup>(yearEventsBox)
                        eventsBox.removeAllViews()

                        var first = true
                        for (event in eventsList) {
                            var title = ""
                            var time = formatTimeStr(requireContext(), event.time, false)
                            var subtitle = ""
                            var link: String? = null
                            var image = 0
                            when (event.type) {
                                EventType.EARTH_APHELION, EventType.EARTH_PERIHELION -> {
                                    if (Prefs.showElement(requireContext(), "yearEarthApsis", true)) {
                                        title = event.type.displayName
                                        link = event.link
                                    }
                                }
                                EventType.PARTIAL_LUNAR, EventType.TOTAL_LUNAR, EventType.PENUMBRAL_LUNAR -> {
                                    if (Prefs.showElement(requireContext(), "yearLunarEclipse", true)) {
                                        title = event.type.displayName
                                        time = "Greatest eclipse: $time"
                                        link = event.link
                                    }
                                }
                                EventType.PARTIAL_SOLAR, EventType.TOTAL_SOLAR, EventType.ANNULAR_SOLAR, EventType.HYBRID_SOLAR -> {
                                    if (!Prefs.showElement(requireContext(), "yearSolarEclipse", true)) {
                                        title = event.type.displayName
                                        time = "Greatest eclipse: $time"
                                        subtitle = event.extra as String
                                        link = event.link
                                    }
                                }
                                EventType.MARCH_EQUINOX, EventType.SEPTEMBER_EQUINOX -> {
                                    if (Prefs.showElement(requireContext(), "yearEquinox", true)) {
                                        title = event.type.displayName
                                    }
                                }
                                EventType.NORTHERN_SOLSTICE -> {
                                    if (Prefs.showElement(requireContext(), "yearSolstice", true)) {
                                        title = event.type.displayName
                                        if (abs(location.location.latitude.doubleValue) > 23.44) {
                                            val sunDay = SunCalculator.calcDay(location.location, event.time, SunCalculator.Event.RISESET)
                                            val localExtreme = if (location.location.latitude.doubleValue >= 0) "Longest" else "Shortest"
                                            subtitle = localExtreme + " day: " + formatDurationHMS(requireContext(), sunDay.uptimeHours, true)
                                        }
                                    }
                                }
                                EventType.SOUTHERN_SOLSTICE -> {
                                    if (Prefs.showElement(requireContext(), "yearSolstice", true)) {
                                        title = event.type.displayName
                                        if (abs(location.location.latitude.doubleValue) > 23.44) {
                                            val sunDay = SunCalculator.calcDay(location.location, event.time, SunCalculator.Event.RISESET)
                                            val localExtreme = if (location.location.latitude.doubleValue >= 0) "Shortest" else "Longest"
                                            subtitle = localExtreme + " day: " + formatDurationHMS(requireContext(), sunDay.uptimeHours, true)
                                        }
                                    }
                                }
                                EventType.PHASE -> {
                                    val moonPhase = event.extra as MoonPhaseEvent?
                                    when (moonPhase?.phase) {
                                        MoonPhase.FULL -> {
                                            if (Prefs.showElement(requireContext(), "yearFullMoon", true)) {
                                                title = "Full Moon"
                                                image = getPhaseFull()
                                            }
                                        }
                                        MoonPhase.NEW -> {
                                            if (Prefs.showElement(requireContext(), "yearNewMoon", true)) {
                                                title = "New Moon"
                                                image = getPhaseNew()
                                            }
                                        }
                                        MoonPhase.FIRST_QUARTER -> {
                                            if (Prefs.showElement(requireContext(), "yearQuarterMoon", true)) {
                                                title = "First Quarter"
                                                image = if (location.location.latitude.doubleValue >= 0) getPhaseRight() else getPhaseLeft()
                                            }
                                        }
                                        MoonPhase.LAST_QUARTER -> {
                                            if (Prefs.showElement(requireContext(), "yearQuarterMoon", true)) {
                                                title = "Last Quarter"
                                                image = if (location.location.latitude.doubleValue >= 0) getPhaseLeft() else getPhaseRight()
                                            }
                                        }
                                        else -> { }
                                    }
                                }
                            }

                            if (title.isNotEmpty()) {
                                if (!first) {
                                    requireActivity().layoutInflater.inflate(R.layout.divider_background, eventsBox)
                                }
                                val eventRow = inflate(R.layout.frag_data_yearevents_event)

                                val today = todayCalendar.get(YEAR) == event.time.get(YEAR) &&
                                        todayCalendar.get(MONTH) == event.time.get(MONTH) &&
                                        todayCalendar.get(DAY_OF_MONTH) == event.time.get(DAY_OF_MONTH)
                                if (today) {
                                    eventRow.setBackgroundColor(getCalendarGridHighlightColor())
                                    eventRow.findViewById<View>(yearEventDateMonth).setBackgroundColor(getCalendarGridHighlightColor())
                                } else {
                                    eventRow.setBackgroundColor(getCalendarDefaultColor())
                                    eventRow.findViewById<View>(yearEventDateMonth).setBackgroundColor(getCalendarHeaderColor())
                                }

                                if (image > 0) {
                                    modifyChild(eventRow, yearEventImg, visibility = VISIBLE, image = image)
                                }
                                modifyChild(eventRow, yearEventDate, visibility = VISIBLE, text = event.time.get(DAY_OF_MONTH).toString())
                                modifyChild(eventRow, yearEventMonth, visibility = VISIBLE, text = getShortMonth(event.time).toUpperCase(Locale.getDefault()))
                                modifyChild(eventRow, yearEventTitle, visibility = VISIBLE, html = title)
                                modifyChild(eventRow, yearEventTime, visibility = VISIBLE, html = time)
                                if (isNotEmpty(subtitle)) {
                                    modifyChild(eventRow, yearEventSubtitle, visibility = VISIBLE, html = subtitle)
                                }
                                if (isNotEmpty(link)) {
                                    modifyChild(eventRow, yearEventLink, visibility = VISIBLE)
                                    eventRow.setOnClickListener {
                                        browseTo(link)
                                    }
                                } else {
                                    eventRow.isClickable = false
                                    eventRow.isFocusable = false
                                }
                                eventsBox.addView(eventRow)
                                first = false
                            }
                        }
                        if (first) {
                            modifyChild(view, yearEventsBox, visibility = GONE)
                            modifyChild(view, yearEventsNone, visibility = VISIBLE)
                        } else {
                            modifyChild(view, yearEventsBox, visibility = VISIBLE)
                            modifyChild(view, yearEventsNone, visibility = GONE)
                        }
                    }
                }
        )
    }

}